### PR TITLE
Use Sources and Sinks for all IO.

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #1241.

### CODE REVIEW

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 1]
```diff
@@ -657,21 +654,23 @@ private void handleWebSocketUpgrade(Socket socket, InputStream in, OutputStream
...
     List<String> headers = response.getHeaders();
     for (int i = 0, size = headers.size(); i < size; i++) {
-      String header = headers.get(i);
-      out.write((header + "\r\n").getBytes(Util.US_ASCII));
+      sink.writeUtf8(headers.get(i));
+      sink.writeUtf8("\r\n");
     }
-    out.write(("\r\n").getBytes(Util.US_ASCII));
-    out.flush();
+    sink.writeUtf8("\r\n");
```
Comment (adriancole) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: like. less cluttery to read.

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 2]
```diff
@@ -756,20 +740,36 @@ public void setDispatcher(Dispatcher dispatcher) {
     return "MockWebServer[" + port + "]";
   }
 
-  /** An output stream that drops data after bodyLimit bytes. */
```
Comment (adriancole) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: accident, or just saying this is intuitive?

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 3]
```diff
@@ -756,20 +740,36 @@ public void setDispatcher(Dispatcher dispatcher) {
...
-    private long numBytesReceived = 0;
+  private static class TruncatingSink implements Sink {
+    private final Buffer buffer = new Buffer();
+    private long remainingByteCount;
+    private long receivedByteCount;
 
-    @Override public void write(byte[] buffer, int offset, int len) {
-      numBytesReceived += len;
-      super.write(buffer, offset, Math.min(len, bodyLimit - count));
+    TruncatingSink(long bodyLimit) {
```
Comment (adriancole) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: maybe the right comment is: Sink that skips after reading {@code bodyLimit} from any source.

**Comment by adriancole** [comment]
LGTM

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 4]
```diff
@@ -522,7 +522,7 @@ private RecordedRequest readRequest(Socket socket, InputStream in, OutputStream
     boolean chunked = false;
     boolean expectContinue = false;
     String header;
-    while ((header = readAsciiUntilCrlf(in)).length() != 0) {
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: Glad to see this go!

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 5]
```diff
@@ -581,20 +581,17 @@ private RecordedRequest readRequest(Socket socket, InputStream in, OutputStream
       throw new UnsupportedOperationException("Unexpected method: " + request);
     }
 
-    return new RecordedRequest(request, headers, chunkSizes, requestBody.numBytesReceived,
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: how did a `num` prefix sneak in here? I don't do that!

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 6]
```diff
@@ -859,30 +859,9 @@ private void writeResponse(SpdyStream stream, MockResponse response) throws IOEx
...
-          } catch (InterruptedException e) {
-            throw new AssertionError(e);
-          }
-        }
         BufferedSink sink = Okio.buffer(stream.getSink());
-        if (response.getThrottleBytesPerPeriod() == Integer.MAX_VALUE) {
-          sink.writeAll(body);
-          sink.flush();
-        } else {
-          while (body.size() > 0) {
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: Winning.

**Comment by swankjesse** [comment]
LGTM

**Pull request merged**